pkg/controller/job/controller.go

```diff
@@ -324,12 +327,47 @@ func (jm *JobController) syncJob(key string) error {
 	activePods := controller.FilterActivePods(podList.Items)
 	active := len(activePods)
 	succeeded, failed := getStatus(podList.Items)
-	if jobNeedsSync {
-		active = jm.manageJob(activePods, succeeded, &job)
+	if job.Status.StartTime == nil {
+		now := unversioned.Now()
+		job.Status.StartTime = &now
 	}
-	completions := succeeded
-	if completions == *job.Spec.Completions {
-		job.Status.Conditions = append(job.Status.Conditions, newCondition())
+	if pastActiveDeadline(&job) {
+		// if job was finished previously, we don't want to redo the termination
+		if isJobFinished(&job) {
+			return nil
+		}
+		// TODO: below code should be replaced with pod termination resulting in
+		// pod failures, rather than killing pods. Unfortunately none such solution
+		// exists ATM. There's an open discussion in the topic in
+		// https://github.com/kubernetes/kubernetes/issues/14602 which might give
+		// some sort of solution to above problem.
+		// kill remaining active pods
+		wait := sync.WaitGroup{}
+		wait.Add(active)
+		for i := 0; i < active; i++ {
+			go func(ix int) {
+				defer wait.Done()
+				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name); err != nil {
+					defer util.HandleError(err)
+				}
+			}(i)
+		}
+		wait.Wait()
+		// update status values accordingly
+		failed += active
+		active = 0
+		job.Status.Conditions = append(job.Status.Conditions, newCondition(extensions.JobFail, "DeadlineExceeded", "Job was active longer than specified deadline"))
+		jm.recorder.Event(&job, api.EventTypeNormal, "DeadlineExceeded", "Job was active longer than specified deadline")
+	} else {
+		if jobNeedsSync {
+			active = jm.manageJob(activePods, succeeded, &job)
+		}
+		completions := succeeded
+		if completions == *job.Spec.Completions {
+			job.Status.Conditions = append(job.Status.Conditions, newCondition(extensions.JobComplete, "", ""))
+			now := unversioned.Now()
+			job.Status.CompletionTime = &now
+		}
 	}
 
 	// no need to update the job if the status hasn't changed since last time
```

Comment:
If no pods of the job have been started yet, and the deadline is already exceeded, then the condition on the next line will be false, the updateHandler won't ever run, so the updated job.Status.Conditions will not be posted.  I think.